subprojects/core/src/main/java/org/gradle/api/internal/initialization/ResettableConfiguration.java

```diff
@@ -47,6 +52,7 @@ public interface ResettableConfiguration extends Configuration {
      * @implSpec Usage: This method should only be called on resolvable configurations and should throw an exception if
      * called on a configuration that does not permit this usage.
      */
-    void resetResolutionState();
+    @Nullable
+    <T> T withResetResolutionState(Factory<T> factory);
```

Comment:
Changing this method to accept a factory seems more complicated than I would have thought necessary. 

Why not just call `setResolutionShouldBeFinal(false)` specifically for the `classpath` configuration when creating that configuration? Then I don't think this method would need to change at all

Reply:
Per our conversation, this is partly because `ResolutionStrategyInternal` is not visible from `core` and partly so that we don't have to always manage both things when resetting the resolution state - instead, there is just one method that takes care of everything properly.